### PR TITLE
Fix destruction order of Suggest #16035

### LINKED ISSUE #16035: clickhouse-client and 100K tables crashes on exit
Related to huge suggestions list

```
(gdb) thread apply all bt

Thread 2 (Thread 0x7fbefa6d8400 (LWP 298)):
#0  __pthread_clockjoin_ex (threadid=140458186086144, thread_return=0x0, clockid=<optimized out>, abstime=<optimized out>, block=<optimized out>) at pthread_join_common.c:145
#1  0x0000000007c42cb0 in DB::Suggest::~Suggest() ()
#2  0x00007fbefa874a27 in __run_exit_handlers (status=0, listp=0x7fbefaa16718 <__exit_funcs>, run_list_atexit=run_list_atexit@entry=true, run_dtors=run_dtors@entry=true) at exit.c:108
#3  0x00007fbefa874be0 in __GI_exit (status=<optimized out>) at exit.c:139
#4  0x00007fbefa8520ba in __libc_start_main (main=0x7b8e780 <main>, argc=1, argv=0x7ffc4e5be1e8, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7ffc4e5be1d8) at ../csu/libc-start.c:342
#5  0x0000000007b3f02e in _start ()

Thread 1 (Thread 0x7fbef8499700 (LWP 299)):
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007fbefa850859 in __GI_abort () at abort.c:79
#2  0x0000000010eb0edc in Poco::SignalHandler::handleSignal(int) ()
#3  <signal handler called>
#4  0x000000000d42dd4f in std::__1::__hash_const_iterator<std::__1::__hash_node<std::__1::__hash_value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, void*>*> std::__1::__hash_table<std::__1::__hash_value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::__unordered_map_hasher<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::__hash_value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, true>, std::__1::__unordered_map_equal<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::__hash_value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, true>, std::__1::allocator<std::__1::__hash_value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > >::find<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) const ()
#5  0x000000000d793065 in DB::IFactoryWithAliases<std::__1::function<std::__1::shared_ptr<DB::IDataType const> (std::__1::shared_ptr<DB::IAST> const&)> >::getAliasToOrName(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) const ()
#6  0x000000000d7927f1 in DB::DataTypeFactory::get(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::IAST> const&) const ()
#7  0x000000000d792640 in DB::DataTypeFactory::get(std::__1::shared_ptr<DB::IAST> const&) const ()
#8  0x000000000d7924b1 in DB::DataTypeFactory::get(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) const ()
#9  0x000000000ddfd267 in DB::NativeBlockInputStream::readImpl() ()
#10 0x000000000d7575d5 in DB::IBlockInputStream::read() ()
#11 0x000000000e530096 in DB::Connection::receivePacket() ()
#12 0x0000000007c6e217 in DB::Suggest::fetch(DB::Connection&, DB::ConnectionTimeouts const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) ()
#13 0x0000000007c6e0a4 in DB::Suggest::loadImpl(DB::Connection&, DB::ConnectionTimeouts const&, unsigned long) ()
#14 0x0000000007c6eb14 in ?? ()
#15 0x00007fbefaa37609 in start_thread (arg=<optimized out>) at pthread_create.c:477
#16 0x00007fbefa94d293 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

### PR DESCRIPTION
**Pull request by alexey-milovidov**
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix rare issue when clickhouse-client may abort on exit due to loading of suggestions. This fixes #16035.

### CODE REVIEW

File: tests/queries/0_stateless/01526_client_start_and_exit.sh [hunk 1]
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/env bash
+
+CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+. "$CURDIR"/../shell_config.sh
+
+# Create a huge amount of tables, so Suggest will take a time to load
+seq 1 1000 | sed -r -e 's/(.+)/CREATE TABLE IF NOT EXISTS test\1 (x UInt8) ENGINE = Memory;/' | ${CLICKHOUSE_CLIENT} -n
```
Comment (filimonov) [inline tests/queries/0_stateless/01526_client_start_and_exit.sh:7]: maybe try longer / random-like table/column names?

File: tests/queries/0_stateless/01526_client_start_and_exit.expect [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+log_user 0
+set timeout 5
+match_max 100000
+
+if ![info exists env(CLICKHOUSE_PORT_TCP)] {set env(CLICKHOUSE_PORT_TCP) 9000}
+
+spawn clickhouse-client --port "$env(CLICKHOUSE_PORT_TCP)"
+expect ":) "
+send -- "\4"
```
Comment (filimonov) [inline tests/queries/0_stateless/01526_client_start_and_exit.expect:11]: maybe add comment: "\4 = EOT (end of transmission), corresponds to pressing Ctrl+D"

File: tests/queries/0_stateless/01526_client_start_and_exit.sh (hunk 1 above)
Comment (filimonov) [reply tests/queries/0_stateless/01526_client_start_and_exit.sh:7]: Checked, it didn't help.

It seems it start reporducing on really big number of tables (maybe it makes SELECT which feed the suggestion list slow?)

Check this out:
```
root@27e49329c2e8:/# clickhouse-client -q "SELECT 'CREATE TABLE test_' || hex(randomPrintableASCII(40)) || '(x UInt8) Engine=Memory;' from numbers(200000)" --format=TSVRaw | clickhouse-benchmark -c32 -i 200000 -d 0
Loaded 200000 queries.

Queries executed: 200000 (100.000%).

localhost:9000, queries 200000, QPS: 3528.722, RPS: 0.000, MiB/s: 0.000, result RPS: 0.000, result MiB/s: 0.000.

0.000%		0.001 sec.	
10.000%		0.002 sec.	
20.000%		0.002 sec.	
30.000%		0.002 sec.	
40.000%		0.003 sec.	
50.000%		0.003 sec.	
60.000%		0.003 sec.	
70.000%		0.004 sec.	
80.000%		0.004 sec.	
90.000%		0.026 sec.	
95.000%		0.053 sec.	
99.000%		0.089 sec.	
99.900%		0.205 sec.	
99.990%		0.346 sec.	


root@27e49329c2e8:/# clickhouse-client 
ClickHouse client version 20.9.3.45 (official build).
Connecting to localhost:9000 as user default.
Connected to ClickHouse server version 20.9.3 revision 54439.

27e49329c2e8 :) Bye.
Aborted (core dumped)
root@27e49329c2e8:/# clickhouse-client 
ClickHouse client version 20.9.3.45 (official build).
Connecting to localhost:9000 as user default.
Connected to ClickHouse server version 20.9.3 revision 54439.

27e49329c2e8 :) Bye.
Aborted (core dumped)
root@27e49329c2e8:/# clickhouse-client 
ClickHouse client version 20.9.3.45 (official build).
Connecting to localhost:9000 as user default.
Connected to ClickHouse server version 20.9.3 revision 54439.

27e49329c2e8 :) Bye.
Aborted (core dumped)
root@27e49329c2e8:/# clickhouse-client 
ClickHouse client version 20.9.3.45 (official build).
Connecting to localhost:9000 as user default.
Connected to ClickHouse server version 20.9.3 revision 54439.

27e49329c2e8 :) Bye.
Aborted (core dumped)
```

Comment (filimonov) [reply tests/queries/0_stateless/01526_client_start_and_exit.sh:7]: In my case it seems it starts reproducing when you have > 20K tables, and reproduces quite stable when I have about 50K of tables and hit Ctrl+D fast enough.

```
clickhouse-client -q "SELECT 'CREATE TABLE test_' || toString(number) || '(x UInt8) Engine=Memory;' from numbers(50000)" --format=TSVRaw | clickhouse-benchmark -c32 -i 50000 -d 0
```

File: tests/queries/0_stateless/01526_client_start_and_exit.expect (hunk 2 above)
Comment (filimonov) [inline tests/queries/0_stateless/01526_client_start_and_exit.expect:11]:
```suggestion
send -- "\4" # EOT (end of transmission), corresponds to pressing Ctrl+D to leave client
```

File: tests/queries/0_stateless/01526_client_start_and_exit.sh (hunk 1 above)
Comment (alexey-milovidov) [reply tests/queries/0_stateless/01526_client_start_and_exit.sh:7]: @filimonov You did not test carefully*,
this patch fixes the issue.

\* maybe forgot to replace `clickhouse-client` to the new build.

Comment (filimonov) [reply tests/queries/0_stateless/01526_client_start_and_exit.sh]: Of course, i was checking a build w/o the fix applied - you just was saying you were not able to reproduce the problem, so i tried to explain a reproduce scenario better.

Fix looks good, and definitely should solve the issue.

**Pull request merged**